components/feature/downloads/src/main/java/mozilla/components/feature/downloads/share/ShareDownloadMiddleware.kt

```
@@ -0,0 +1,180 @@
+ *
+ * Following which it will transparently
+ *  - download internet resources while respecting the private mode related to cookies handling
+ *  - temporarily cache the downloaded resources
+ *  - automatically open the platform app chooser to share the cached files with other installed Android apps.
+ *
+ * with a 1 second timeout to ensure a smooth UX.
+ *
+ *  @param context Android context used for vaious platform interactions
+ *  @param httpClient Client used for downloading internet resources
+ *  @param scope CoroutineScope used for background work execution
+ */
+class ShareDownloadMiddleware(
+    private val context: Context,
+    private val httpClient: Client,
+    private val scope: CoroutineScope = CoroutineScope(IO)
+) : Middleware<BrowserState, BrowserAction> {
+    @VisibleForTesting
+    internal val logger = Logger("ShareDownloadMiddleware")
+
+    init {
+        scope.launch {
+            cleanupCache()
+        }
+    }
+
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        if (action is DownloadAction.AddDownloadAction &&
+            action.download.source == DownloadState.Source.SHARE
+        ) {
+            logger.debug("Starting the sharing process")
+            startSharing(action.download)
+        } else {
+            next(action)
+        }
+    }
+
+    @VisibleForTesting
+    internal fun startSharing(fileToShare: DownloadState) {
+        val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
+            when (throwable) {
+                is InterruptedException -> {
+                    logger.warn("Share failed: operation timeout reached")
+                }
+                is IOException,
+                is RuntimeException,
+                is NullPointerException -> {
+                    logger.warn("Share failed: $throwable")
+                }
```

Should we should some UI to inform the user that something unexpected happened while sharing?

---

When downloading we use a toast to make the user aware of some errors https://github.com/mozilla-mobile/android-components/blob/c1a85a4254a7a79eab60fe3c71a67e15f7db101f/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L232

---

Saw Chrome failing silently but maybe we could do better.
Asked UX in the ticket.

---

👍🏽